grove/simon/simon.py

```diff
@@ -0,0 +1,450 @@
+    :math:`f:\\{0,1\\}^n\\rightarrow \\{0,1\\}^n`, such that
+
+    .. math::
+
+        U_f\\vert x\\rangle = \\vert f(x)\\rangle
+
+    create an oracle program that performs the following transformation:
+
+    .. math::
+
+        \\vert x \\rangle \\vert y \\rangle
+        \\rightarrow \\vert x \\rangle \\vert f(x) \\oplus y\\rangle
+
+    where :math:`\\vert x\\rangle` and :math:`\\vert y\\rangle`
+    are :math:`n` qubit states and :math:`\\oplus` is bitwise xor.
+
+    Allocates one scratch bit.
+
+    :param 2darray unitary_funct: Matrix representation :math:`U_f` of the
+                                  function :math:`f`,
+                                  i.e. the unitary transformation
+                                  that must be applied to a state
+                                  :math:`\\vert x \\rangle`
+                                  to get :math:`\\vert f(x) \\rangle`
+    :param list(int) qubits: List of qubits that enter as the input
+                        :math:`\\vert x \\rangle`.
+    :param list(int) ancillas: List of qubits to serve as the ancillary input
+                     :math:`\\vert y \\rangle`.
+    :return: A program that performs the above unitary transformation.
+    :rtype: Program
+    """
+    assert is_unitary(unitary_funct), "Function must be unitary."
+    p = pq.Program()
+
+    scratch_bit = p.alloc()
+    bits_for_funct = [scratch_bit] + qubits
+
+    p.defgate("FUNCT", unitary_funct)
+    p.defgate("FUNCT-INV", np.linalg.inv(unitary_funct))
+    p.inst(tuple(['FUNCT'] + bits_for_funct))
+    p.inst(map(lambda qs: CNOT(qs[0], qs[1]), zip(qubits, ancillas)))
+    p.inst(tuple(['FUNCT-INV'] + bits_for_funct))
+
+    p.free(scratch_bit)
+    return p
+
+
+def simon(oracle, qubits):
+    """
+    Implementation of the quantum portion of Simon's Algorithm.
+
+    Given a list of input qubits,
+    all initially in the :math:`\\vert 0\\rangle` state,
+    create a program that applies the Hadamard-Walsh transform the qubits
+    before and after going through the oracle.
+
+    :param Program oracle: Program representing unitary application of function
+    :param list(int) qubits: List of qubits that enter as the input
+                        :math:`\\vert x \\rangle`.
+    :return: A program corresponding to the desired instance of
+             Simon's Algorithm.
+    :rtype: Program
+    """
+    p = pq.Program()
+
+    # Apply Hadamard, Unitary function, and Hadamard again
+    p.inst(map(H, qubits))
+    p += oracle
+    p.inst(map(H, qubits))
+    return p
+
+
+def is_unitary(matrix):
+    """
+    A helper function that checks if a matrix is unitary.
+
+    :param 2darray matrix: a matrix to test unitarity of
+    :return: true if and only if matrix is unitary
+    :rtype: bool
+    """
+    rows, cols = matrix.shape
+    if rows != cols:
+        return False
+    return np.allclose(np.eye(rows), matrix.dot(matrix.T.conj()))
+
+
+def most_significant_bit(lst):
```

Comment:
Though I'm not convinced that the documentation is correct. You say `lst` is a list (in which case `index` will work), but you use `most_significant_bit` below on an `ndarray`. If it should work for ndarrays, you should change the documentation.